File: analytics/src/main/java/io/appmetrica/analytics/impl/crash/ReadAndReportRunnable.java
package io.appmetrica.analytics.impl.crash;

import androidx.annotation.NonNull;
import io.appmetrica.analytics.coreapi.internal.backport.Consumer;
import io.appmetrica.analytics.coreapi.internal.backport.Function;
import io.appmetrica.analytics.coreutils.internal.logger.YLogger;
import java.io.File;

public class ReadAndReportRunnable<Output> implements Runnable {

    private static final String TAG = "[ReadAndReportRunnable]";

    @NonNull
    private final File crashFile;
    @NonNull
    private final Function<File, Output> fileReader;
    @NonNull
    private final Consumer<File> finalizator;
    @NonNull
    private final Consumer<Output> crashConsumer;

    public ReadAndReportRunnable(
            @NonNull File crashFile,
            @NonNull Function<File, Output> fileReader,
            @NonNull Consumer<File> finalizator,
            @NonNull Consumer<Output> crashConsumer
    ) {
        this.crashFile = crashFile;
        this.fileReader = fileReader;
        this.finalizator = finalizator;
        this.crashConsumer = crashConsumer;
    }

    @Override
    public void run() {
        if (crashFile.exists()) {
            try {
                Output result = fileReader.apply(crashFile);
                if (result != null) {
                    YLogger.debug(TAG, "for file %s result is %s", crashFile.getName(), result);
                    crashConsumer.consume(result);
                } else {
                    YLogger.debug(TAG, "for file %s result is null", crashFile.getName());
                }
            } catch (Throwable exception) {
                YLogger.error(TAG, exception, "can't handle crash in file %s due to Exception", crashFile.getName());
            } finally {
                finalizator.consume(crashFile);
            }
        }
    }
}

File: analytics/src/main/java/io/appmetrica/analytics/impl/crash/ReadAndReportRunnable.kt
package io.appmetrica.analytics.impl.crash

import io.appmetrica.analytics.coreapi.internal.backport.Consumer
import io.appmetrica.analytics.coreapi.internal.backport.Function
import io.appmetrica.analytics.impl.crash.service.ShouldSendCrashNowPredicate
import io.appmetrica.analytics.impl.utils.concurrency.FileLocksHolder
import io.appmetrica.analytics.logger.appmetrica.internal.DebugLogger
import java.io.File

internal class ReadAndReportRunnable<Output>(
    private val crashFile: File,
    private val fileReader: Function<File, Output?>,
    private val finalizator: Consumer<File>,
    private val crashConsumer: Consumer<Output>,
    private val fileLocksHolder: FileLocksHolder,
    private val shouldSendCrashNowPredicate: ShouldSendCrashNowPredicate<Output>
) : Runnable {

    private val tag = "[ReadAndReportRunnable]"

    override fun run() {
        if (crashFile.exists()) {
            val fileLock = fileLocksHolder.getOrCreate(crashFile.name)
            var postProcessAction = finalizator
            try {
                fileLock.lock()
                if (!crashFile.exists()) {
                    return // crash was processed from another thread
                }
                val result = fileReader.apply(crashFile)
                if (result != null) {
                    if (shouldSendCrashNowPredicate.shouldSend(result)) {
                        DebugLogger.info(tag, "for file ${crashFile.name} result is $result")
                        crashConsumer.consume(result)
                    } else {
                        DebugLogger.info(tag, "for file ${crashFile.name} sending forbid by should send predicate")
                        postProcessAction = Consumer {}
                    }
                } else {
                    DebugLogger.info(tag, "for file ${crashFile.name} result is null")
                }
            } catch (exception: Throwable) {
                DebugLogger.error(
                    tag,
                    exception,
                    "can't handle crash in file ${crashFile.name} due to Exception"
                )
            } finally {
                postProcessAction.consume(crashFile)
                fileLock.unlockAndClear()
                fileLocksHolder.clear(crashFile.name)
            }
        }
    }
}
